### PR TITLE
 Extend SliderBar with PgUp/PgDn/Home/End handling

### PR DESCRIPTION
**Pull request by UselessToucan**
We might want to handle PgUp/PgDn/Home/End in addition to Right/Left arrows.

### CODE REVIEW

**Comment by peppy** [comment]
I dunno. Home and end maybe, but pgup/pgdn should be reserved for vertical scrolling IMO.

- Windows doesn't handle them on slider bars.
- macOS does, but not in a way any user would expect.

Proposal: ctrl+left/right arrows for larger step instead.

**Comment by swoolcock** [comment]
Note that on macOS Cmd+Up/Cmd+Down is the usual combo for "go to top" and "go to bottom".
You may want to use `PlatformAction`s instead.

**Comment by UselessToucan** [comment]
Not sure that I understood change requests correctly, but PlatformAction handling was added.
Btw, it does not work correctly whenever we have TextBoxes with HoldFocus = true(Song select, music controller, settings).

File: osu.Framework/Graphics/UserInterface/SliderBar.cs [hunk 1]
```diff
@@ -108,26 +125,56 @@ protected override bool OnDragStart(InputState state)
...
-                    return true;
-                case Key.Left:
-                    CurrentNumber.Add(-step);
-                    OnUserChange();
+                case PlatformActionType.CharPrevious:
+                case PlatformActionType.CharNext:
+                case PlatformActionType.WordPrevious:
+                case PlatformActionType.WordNext:
+                case PlatformActionType.LineStart:
+                case PlatformActionType.LineEnd:
```
Comment (smoogipoo) [inline osu.Framework/Graphics/UserInterface/SliderBar.cs]: I've never seen a sliderbar that handles home/end. I can't immediately think off the top of my head for a reason to want to jump straight to the extremums of a sliderbar - typically you want to adjust them in small increments.

I think we should leave this out of the framework implementation, and if it ever comes up implement it locally where required.

File: osu.Framework/Graphics/UserInterface/SliderBar.cs [hunk 2]
```diff
@@ -108,26 +125,56 @@ protected override bool OnDragStart(InputState state)
...
                     return true;
-                default:
-                    return false;
+                default: return false;
```
Comment (smoogipoo) [inline osu.Framework/Graphics/UserInterface/SliderBar.cs]: return on a new line please

File: osu.Framework/Graphics/UserInterface/SliderBar.cs [hunk 3]
```diff
@@ -108,26 +125,56 @@ protected override bool OnDragStart(InputState state)
...
-                    CurrentNumber.Add(step);
-                    OnUserChange();
-                    return true;
-                case Key.Left:
-                    CurrentNumber.Add(-step);
-                    OnUserChange();
+                case PlatformActionType.CharPrevious:
+                case PlatformActionType.CharNext:
+                case PlatformActionType.WordPrevious:
+                case PlatformActionType.WordNext:
```
Comment (smoogipoo) [inline osu.Framework/Graphics/UserInterface/SliderBar.cs]: We shouldn't be reusing action types like this - I think these should be separate to text control.

File: osu.Framework/Graphics/UserInterface/SliderBar.cs [hunk 4]
```diff
@@ -24,10 +24,27 @@ public abstract class SliderBar<T> : Container, IHasCurrentValue<T>
...
+        public float SmallKeyboardStep
+        {
+            get => smallKeyboardStep != default(float) ? smallKeyboardStep : defaultSmallKeyboardStep;
+            set => smallKeyboardStep = value;
+        }
+
+        private float largeKeyboardStep;
+        private float defaultLargeKeyboardStep => (Convert.ToSingle(CurrentNumber.MaxValue) - Convert.ToSingle(CurrentNumber.MinValue)) / 4;
+        /// <summary>
+        /// A custom step value for each PgUp/PgDn key press which actuates a change on this control.
```
Comment (smoogipoo) [inline osu.Framework/Graphics/UserInterface/SliderBar.cs]: XMLDoc is outdated

File: osu.Framework/Graphics/UserInterface/SliderBar.cs [hunk 5]
```diff
@@ -24,10 +24,27 @@ public abstract class SliderBar<T> : Container, IHasCurrentValue<T>
...
         /// </summary>
-        public float KeyboardStep;
+        public float SmallKeyboardStep
+        {
+            get => smallKeyboardStep != default(float) ? smallKeyboardStep : defaultSmallKeyboardStep;
+            set => smallKeyboardStep = value;
+        }
+
+        private float largeKeyboardStep;
+        private float defaultLargeKeyboardStep => (Convert.ToSingle(CurrentNumber.MaxValue) - Convert.ToSingle(CurrentNumber.MinValue)) / 4;
```
Comment (smoogipoo) [inline osu.Framework/Graphics/UserInterface/SliderBar.cs]: Does this even work with, e.g. an int sliderbar with range [4, 5]?

File: osu.Framework/Graphics/UserInterface/SliderBar.cs (hunk 4 above)
Comment (UselessToucan) [reply osu.Framework/Graphics/UserInterface/SliderBar.cs]: I don't understand what is wrong

**Pull request closed without merging**